### PR TITLE
add support for JSON parameters on win

### PR DESCRIPTION
**Pull request by toddmedema**
Closes https://github.com/Fabricate-IO/expedition-app/issues/76

Implemented in such a manner that it'd (hopefully) be easy to extend for other future triggers

### CODE REVIEW

**Comment by toddmedema** [comment]
Follows the discussion in the issue, but on the event trigger instead of general combat (allowing for more flexibility, since different combat outcomes could result in no loot or no healing, etc)

ie `* on win {"loot": false, "exp": false, "heal": 2}`

Getting tests working is out of the scope of this PR, but added a note to the combat test so that we remember to test it when we get them working again

**Comment by toddmedema** [comment]
Added a QC issue to support the rendering of trigger JSON: https://github.com/Fabricate-IO/expedition-quest-creator/issues/197

File: app/QuestParser.tsx [hunk 1]
```diff
@@ -68,6 +69,32 @@ export function handleEvent(parent: XMLElement, event: string, ctx: QuestContext
   if (!child) {
     throw new Error('Could not find child with on="'+event+'"');
   }
+  return child;
+};
+
+// The passed event parameter is a string indicating which event to fire based on the "on" attribute.
+// Returns the (cleaned) parameters of the event element
+export function getEventParameters(parent: XMLElement, event: string, ctx: QuestContext): any {
```
Comment (smartin015) [inline app/QuestParser.tsx]: Suggest doing per-attribute validation instead of inferring values.

File: app/components/Combat.test.tsx [hunk 2]
```diff
@@ -172,4 +172,7 @@ test('end combat stats should be non-NaN', function() {
   throw new Error('unimplemented');
 });
 
+TODO test combat trigger JSON params, ie * on win {"loot": false, "exp": false, "heal": 2}
```
Comment (smartin015) [inline app/components/Combat.test.tsx]: Can you put this in jasmine format at least?

File: app/components/Combat.tsx [hunk 3]
```diff
@@ -209,20 +210,19 @@ function renderPlayerTier(props: CombatProps): JSX.Element {
...
   var contents: JSX.Element[] = [];
-
-  if (props.settings.showHelp) {
-    contents.push(
-      <p key="c1">All adventurers (dead and alive) heal to full health.
-      </p>
-    );
+  const victoryParameters = getEventParameters(props.node, 'win', props.ctx);
+
+  if (victoryParameters.heal > 0 && victoryParameters.heal < 12) {
```
Comment (smartin015) [inline app/components/Combat.tsx]: Suggest adding `12` to constants (MAX_PLAYER_HEALTH?)

File: app/components/Combat.tsx [hunk 4]
```diff
@@ -209,20 +210,19 @@ function renderPlayerTier(props: CombatProps): JSX.Element {
...
   }
 
-  if (props.combat.levelUp) {
-    contents.push(
-      <p key="c2">
-        Each Adventurer may learn a new ability:
-      </p>
-    );
+  // TODO improved leveling up: https://github.com/Fabricate-IO/expedition-app/issues/226
+  if (victoryParameters.exp !== false) {
```
Comment (smartin015) [inline app/components/Combat.tsx]: Should add levelUp conditional here. Also, exp -> xp

File: app/reducers/QuestTypes.tsx [hunk 5]
```diff
@@ -64,8 +64,9 @@ export interface MidCombatPhase {
   tier: number;
 }
 export interface EndCombatPhase {
-  loot?: Loot[];
+  heal?: number;
```
Comment (smartin015) [inline app/reducers/QuestTypes.tsx]: Where is this used? I don't see `props.combat.heal` anywhere.

File: app/QuestParser.tsx (hunk 1 above)
Comment (toddmedema) [inline app/QuestParser.tsx]: done

File: app/components/Combat.tsx (hunk 4 above)
Comment (toddmedema) [inline app/components/Combat.tsx]: done

File: app/components/Combat.tsx (hunk 3 above)
Comment (toddmedema) [inline app/components/Combat.tsx]: done

File: app/components/Combat.test.tsx (hunk 2 above)
Comment (toddmedema) [inline app/components/Combat.test.tsx]: done

File: app/reducers/QuestTypes.tsx (hunk 5 above)
Comment (toddmedema) [inline app/reducers/QuestTypes.tsx]: Ah yep, cruft from an old implementation idea

**Comment by toddmedema** [comment]
@smartin015 done

**Pull request merged**